session_test.go

```diff
@@ -613,3 +613,57 @@ func TestSessionFlowFrameWithEcho(t *testing.T) {
 	time.Sleep(100 * time.Millisecond)
 	require.NoError(t, client.Close())
 }
+
+func TestSessionInvalidAttachDeadlock(t *testing.T) {
+	// https://github.com/Azure/go-amqp/issues/87
+	t.Skip("TODO: deadlock fix for conn and session")
+	var enqueueFrames func(string)
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.PerformOpen("container")
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			enqueueFrames(tt.Name)
+			return nil, nil
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+
+	enqueueFrames = func(n string) {
+		// send an invalid attach response
+		b, err := mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformAttach{
+			Name: "mismatched",
+			Role: encoding.RoleReceiver,
+		})
+		require.NoError(t, err)
+		netConn.SendFrame(b)
+		// now follow up with a detach frame
+		b, err = mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformDetach{
+			Error: &encoding.Error{
+				Condition:   "boom",
+				Description: "failed",
+			},
+		})
+		require.NoError(t, err)
+		netConn.SendFrame(b)
+	}
+	snd, err := session.NewSender()
+	require.Error(t, err)
```

Comment:
Might be useful to check the error message here. I'm assuming it's going be complaining about both frames or something similar to that?

Reply:
Fixed